### ISSUE TITLE
Adapter append/prepend breaks virtualization when eof/bof reached

### ISSUE THREAD

**Issue by DaveLomber** [dhilt/ngx-ui-scroll#114]
Hi there

we are building chat view

we have an array `this.messages` with a list of messages sorted by date sent.

Let's say we have 100 messages loaded.

Here is our code of data source implementation:

    datasource: Datasource;
    MIN_DATASOURCE_ITEMS_INDEX = 1;
    DATA_SOURCE_BUFFER_SIZE = 10;

    initDataSourceIfNull() {

     if (this.datasource) {
       return;
     }

     const startIndex = -5;

     this.datasource = new Datasource ({
      get: (index, count, success) => {
        const data = this.getAdpterData(index, count).reverse();
        success(data);
      },
      settings: {
        startIndex,
        bufferSize: this.DATA_SOURCE_BUFFER_SIZE,
        padding: 1 // to fill N screens in backward and forward directions
      }
     });
    }

    getAdpterData(_index: number, count: number) {
      const index = -_index - count + this.MIN_DATASOURCE_ITEMS_INDEX;
      const data = [];
      const start = Math.max(this.MIN_DATASOURCE_ITEMS_INDEX, index);
      const end = index + count - 1;

      let bof = false;
      if (start <= end) {
        for (let i = start; i <= end; i++) {
          if (!this.messages[i]) {
            bof = true;
            break;
          }
          data.push({...this.messages[i], index: i});
        }
      }

      if (bof) {
        this.getMoreMessages();
      }

      return data;
    }

So far all is good, we can scroll to top and display older messages, with virtualization. 

And now, we have a logic to call `getMoreMessages()` once `bof` is reached and so far this logic works properly.

The real issue comes later, when we obtained new data, e.g. another 100 messages and call 

    doAdapterPrepend(items) {
      const bof = true;
      this.datasource.adapter.prepend(items, bof);
    }

where `items` are 100 new messages, which is already a part of  `this.messages`, so `this.messages` is 200 messages now.

Whatever we use for `bof` (true or false) - it inserts all 100 messages right into DOM, w/o any virtualization. 

What I expect is it will insert like 10 or so and then when we scroll to top - it will call datasource get to insert more.

Could you please let me know what's wrong with this approach

**Comment by DaveLomber** [comment]
btw I can see same effect here in the example

https://dhilt.github.io/ngx-ui-scroll/#/adapter#append-prepend-sync

If I prepend 100, then DOM elements count becomes 115

**Comment by DaveLomber** [comment]
can be easily reproduced here https://stackblitz.com/edit/angular-ngx-ui-scroll-1-2-2-auto-prepend-when-bof

just replace `{{datasource.adapter.version}}` with `{{datasource.adapter.itemsCount}}` and scroll to top

**Comment by dhilt** [comment]
@DaveLomber `Adapter.prepend(items, true)` will virtualize items only if real begin-of-file is reached, when `Adapter.bof` is `true`. I believe `prepend` should be called with `true` second argument in your case. But it's not about the issue you want to avoid.

What if I suggest following solution

 - scroller is full, we are at the top, Adapter.bof is true
 - you retrieve new items from the remote (say 100 items)
 - then you extend your app datasource by new 100 items
 - prepend only 10 new items to the ui-scroll via Adapter
 - reset Adapter.bof flag manually (not implemented)

It may look like `Adapter.reset({ bof: false })`, or even something like `Adapter.consider({ count: 90, backward: true })`. The second call will virtualize 90 items on backward direction based on average item size and reset `Adapter.bof` flag to false. What do you think?

**Comment by DaveLomber** [comment]
> reset Adapter.bof flag manually (not implemented)

that's what I was looking for, actually :) 

If you could add this - it will be super cool.
And your solution looks good, same at my mind

**Comment by DaveLomber** [comment]
As I understand, 

if implement `Adapter.reset({ bof: false })`

we need to reset `buffer.absMinIndex` and `buffer.absMaxIndex` internally

**Comment by dhilt** [comment]
@DaveLomber So do you want to to tell the ui-scroll haw many items the datasource have just received? In that case, I think, we may implement new signature for append/prepend:

```
// pass 10 items, prepend 10 if bof
Adapter.prepend({
  items: [...], // array of 10 items
  bof: true
})
```

```
// pass 10 items, prepend all 10 if bof, virtualize 90 more
Adapter.prepend({
  items: [...], // array of 10 items
  bof: true,
  virtualCount: 90
})
```

```
// pass 100 items, prepend 10 if bof, virtualize other 90
Adapter.prepend({
  items: [...], // array of 100 items
  bof: true,
  count: 10
})
```

```
// pass 0 items, do not prepend anything, but virtualize 100 rows
Adapter.prepend({
  virtualCount: 100
})
```

It will take care of all internal params including bof/eof, min/max indexes etc. I'm just thinking... if you have your own thoughts on the API changes, do not hesitate to put them here too.

**Comment by DaveLomber** [comment]
@dhilt thanks for all the proposed solutions

I just had some tome thinking about it

I prefer do not introduce new API, but instead do some internal changes

e.g. let's say we prepend 100 items

```
Adapter.prepend({
  items: [...], // array of 100 items
  bof: true
})
```

and internally to do the following:

- prepend `settings.bufferSize` items only if bof
- virtualize the rest 
- reset internal `datasource.adapter.bof` -> false, so the further scroll will be initiating `datasource.get` calls

**Comment by dhilt** [comment]
Ok, putting this into the queue...

**Comment by dhilt** [comment]
@DaveLomber I'm going to work on some big changes, so I decided to make an interim release containing the updates that might help you to solve your issue. It's a kind of manual handling of stuff but it might make a life better, also the implementation was cheap enough... The approach is described it PR https://github.com/dhilt/ngx-ui-scroll/pull/119 (especially the last sample, "Prepeding and virtualizing"). The release is already on [npm](https://www.npmjs.com/package/ngx-ui-scroll/v/1.2.3), it is v1.2.3.

So please try v1.2.3 as it might help you to move on. It would be great if you can confirm that it works for you, for I may peacefully switch to another issues this time.

**Comment by DaveLomber** [comment]
@dhilt  thank you so much for your effort!

I will check and let you know

**Comment by DaveLomber** [comment]
There is one issue with 
`this.datasource.adapter.fix({ minIndex: -Infinity})` - it produces an error with warning_1  `it must be an integer`

**Comment by DaveLomber** [comment]
seems an issue with this code:

```
if (type === FixParamType.integer) {
  value = Number(value);
  parsedValue = parseInt(value, 10);
  if (value !== parsedValue) {
    warning_1 = 'it must be an integer';
  }
}
```

I did a quick test and 

`Number(-Infinity)` -> `-Infinity`
`parseInt(Number(-Infinity), 10)` -> `NAN`

so `-Infinity` !== `NAN`

**Comment by DaveLomber** [comment]
If we remove this validation then seems it works pretty good
Can we please fix this?

**Comment by dhilt** [comment]
@DaveLomber I'm so happy you have taken a look at this... That's the reason of why we need to cover ALL the functionality with tests, even if it's related to undocumented/temporary solutions. I made some refactoring and forced myself to add a spec for the Fix process input params: https://github.com/dhilt/ngx-ui-scroll/pull/122. I didn't remove validation, but I took special care of the Infinite cases.

I published the result on npm as v1.2.4
